Add retires for docker-compose pull in integration tests

azat opened the pull request:
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add retires for docker-compose pull in integration tests

Detailed description / Documentation draft:
Recently integration tests stated to fail due to pull failures, let's try 10 retries with 5 second delay (50 seconds in total), in attempt to overcome this issue (another option is to try `--ignore-pull-failures` if there such image already in the local docker cache, but let's try w/o it first)

Cc: @filimonov 
Cc: @alexey-milovidov 
Cc: @qoega

azat committed:
tests/integration: stops orphan contains too

azat committed:
tests/integration: pull images in a separate step with --ignore-pull-failures

azat committed:
tests/integration: drop --force-recreate since we have down/pull separately

azat committed:
tests/integration: add docker-compose pull retries

filimonov commented on tests/integration/helpers/cluster.py:
```diff
@@ -606,13 +626,16 @@ def start(self, destroy_dirs=True):
-            if not subprocess_call(['docker-compose', 'kill']):
-                subprocess_call(['docker-compose', 'down', '--volumes'])
+            subprocess_call(['docker-compose', 'kill'])
+            subprocess_call(['docker-compose', 'down', '--volumes', '--remove-orphans'])
```
Without docker-compose.yml it just produce some eror message

filimonov commented on tests/integration/helpers/cluster.py:
```diff
@@ -624,7 +647,7 @@ def start(self, destroy_dirs=True):
             self.docker_client = docker.from_env(version=self.docker_api_version)
 
-            common_opts = ['up', '-d', '--force-recreate']
+            common_opts = ['up', '-d']
```
Why removing force recreate? I guess it can have a side effects - if test1 start containers and will not remove them, 
test2 can get 'dirty' containers from 1st test run.

qoega replied:
It should be ok inside one integration test file. We just need to add separate names for different cases.

filimonov replied:
Then same problem can happen if the SAME test will be fired twice.

azat replied:
There is `docker-compose down` at the beginning of `start` and it will stop and remove all containers or fail, so `--force-recreate` is not required anymore

qoega commented:
I'll check if it works when we set project name. Or just kill/rm containers and filter by project manually.

Probably it is better to pull all images  before all integration tests run.

azat committed:
tests/integration: run docker-compose commands after copying of compose files

azat commented:
>I'll check if it works when we set project name. Or just kill/rm containers and filter by project manually.

There is already `--project-name` (even default, based on dirname, is ok) - https://github.com/ClickHouse/ClickHouse/blob/ae883829efacd726916fad851fd178f1b32e0a02/tests/integration/helpers/cluster.py#L130

alexey-milovidov commented:
Need to merge with master.
The failure of `test_zookeeper_config` is suspicious.

azat committed:
tests/integration: stop containers at start only if destroy_dirs is not set

Fixes tests with multiple clusters - test_zookeeper_config

azat commented:
>Need to merge with master.

Rebased against upstream/master

>The failure of test_zookeeper_config is suspicious.

Right, multiple clusters had been broken (with correct `docker-compose down`), fixed in https://github.com/ClickHouse/ClickHouse/pull/22647/commits/fad42f8ac5d8f23c2215a84dff8e61be0dc40495

azat commented:
>[Integration tests (release) — fail: 0, passed: 1293, flaky: 1](https://clickhouse-test-reports.s3.yandex.net/22647/fad42f8ac5d8f23c2215a84dff8e61be0dc40495/integration_tests_(release).html)

>[Integration tests (asan) — Timeout, fail: 0, passed: 1290, flaky: 0](https://clickhouse-test-reports.s3.yandex.net/22647/fad42f8ac5d8f23c2215a84dff8e61be0dc40495/integration_tests_(asan).html#fail1)
>[Integration tests (thread) — Timeout, fail: 1, passed: 1231, flaky: 0](https://clickhouse-test-reports.s3.yandex.net/22647/fad42f8ac5d8f23c2215a84dff8e61be0dc40495/integration_tests_(thread).html#fail1)

This happens [from time to time](https://clickhouse-test-reports.s3.yandex.net/22977/0bf6b61b590719e4f0d17dc7b9171d6f8cdd7c29/integration_tests_(thread).html#fail1) in other PRs/commits too.

excitoon commented:
@azat @alexey-milovidov This PR breaks `--disable-net-host` option on any test:

```
Stderr:
Pulling proxy2           ... error
Pulling proxy1           ... error
Pulling minio1           ... error
Pulling s3_max_redirects ... error
Pulling restricted_dummy ... error
Pulling dummy            ... error
Pulling resolver         ... error

ERROR: for minio1  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

ERROR: for dummy  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

ERROR: for resolver  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

ERROR: for s3_max_redirects  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

ERROR: for restricted_dummy  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

ERROR: for proxy1  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

ERROR: for proxy2  Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)


Stdout:


Failed to start cluster: 
Timeout >300.0s
None
----------------------------------------------------------------------- Captured stderr setup ------------------------------------------------------------------------
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...
Starting cluster...

        Can't find a suitable configuration file in this directory or any
        parent. Are you in the right directory?

        Supported filenames: docker-compose.yml, docker-compose.yaml
        
Removing network rootteststorages3_default
Network rootteststorages3_default not found.
Removing volume rootteststorages3_data1-1
Volume rootteststorages3_data1-1 not found.
Traceback (most recent call last):
  File "/ClickHouse/tests/integration/helpers/cluster.py", line 677, in start
    retry_exception(10, 5, subprocess_check_call, Exception, clickhouse_pull_cmd)
  File "/ClickHouse/tests/integration/helpers/cluster.py", line 69, in retry_exception
    func(*args, **kwargs)
  File "/ClickHouse/tests/integration/helpers/cluster.py", line 80, in subprocess_check_call
    run_and_check(args)
  File "/ClickHouse/tests/integration/helpers/cluster.py", line 49, in run_and_check
    res = subprocess.run(args, stdout=stdout, stderr=stderr, env=env, shell=shell)
  File "/usr/lib/python3.6/subprocess.py", line 425, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.6/subprocess.py", line 863, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.6/subprocess.py", line 1534, in _communicate
    ready = selector.select(timeout)
  File "/usr/lib/python3.6/selectors.py", line 376, in select
    fd_event_list = self._poll.poll(timeout)
  File "/usr/local/lib/python3.6/dist-packages/pytest_timeout.py", line 200, in handler
    timeout_sigalrm(item, params.timeout)
  File "/usr/local/lib/python3.6/dist-packages/pytest_timeout.py", line 375, in timeout_sigalrm
    pytest.fail("Timeout >%ss" % timeout)
  File "/usr/local/lib/python3.6/dist-packages/_pytest/outcomes.py", line 153, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: Timeout >300.0s
------------------------------------------------------------------------- Captured log setup -------------------------------------------------------------------------
INFO     root:test.py:93 Starting cluster...
====================================================================== short test summary info =======================================================================
```
